src/main/scala/org/hammerlab/guacamole/reference/ReferenceGenome.scala

```diff
@@ -1,52 +1,147 @@
+          )
+        }
+        val contigLength = pieces(1).toInt
+        val region = LociSet.parse(pieces(0)).result
+        if (region.contigs.length != 1) {
+          throw new IllegalArgumentException(s"Region must have 1 contig for partial fasta: ${pieces(0)}")
+        }
+        val contig = region.contigs.head
+        val regionLength = region.onContig(contig).count
+        if (regionLength != sequence.length) {
+          throw new IllegalArgumentException(
+            "In partial fasta, region %s is length %,d but its sequence is length %,d".format(
+              pieces(0), regionLength, sequence.length
+            )
+          )
+        }
+        val maxRegion = region.onContig(contig).ranges.map(_.end).max
+        if (maxRegion > contigLength) {
+          throw new IllegalArgumentException(
+            "In partial fasta, region %s (max=%,d) exceeds contig length %,d".format(
+              pieces(0), maxRegion, contigLength
+            )
+          )
+        }
+        if (contigLengths.getOrElseUpdate(contig, contigLength) != contigLength) {
+          throw new IllegalArgumentException(
+            "In partial fasta, contig lengths for %s are inconsistent (%d vs %d)".format(
+              contig, contigLength, contigLengths(contig)
+            )
+          )
+        }
+        val sequenceMap = result.getOrElseUpdate(contig, mutable.HashMap[Int, Byte]())
+        region.onContig(contig).iterator.zip(sequence.toIterator).foreach(pair => {
+          sequenceMap.update(pair._1.toInt, pair._2)
+        })
+      }
+    })
+    new ReferenceGenome(
+      result.map(
+        pair =>
+          pair._1 ->
+            MapBackedReferenceSequence(
+              contigLengths(pair._1),
+              pair._2.toMap
+            )
+      ).toMap
+    )
+  }
 
   /**
-   * Build a new MDTag from the reference and read sequence
+   * Load a ReferenceBroadcast, caching the result.
    *
-   * @param readSequence Bases from the read
-   * @param referenceContig Name of the reference contig or chromosome
-   * @param referenceStart Start position of the read on the reference
-   * @param cigar CIGAR for the read on the reference
-   * @return MdTag for the the read
+   * @param fastaPath Local path to a FASTA file
+   * @return ReferenceBroadcast which maps contig/chromosome names to broadcasted sequences
    */
-  def buildMdTag(readSequence: String, referenceContig: String, referenceStart: Int, cigar: Cigar): String = {
-    val referenceSequence =
-      Bases.basesToString(
-        getReferenceSequence(referenceContig, referenceStart, referenceStart + cigar.getReferenceLength)
-      )
-    MdTag(readSequence, referenceSequence, cigar, referenceStart).toString
+  def apply(fastaPath: String, partialFasta: Boolean = false): ReferenceGenome = {
+    cache.getOrElseUpdate(
+      fastaPath,
+      if (partialFasta)
```

Comment:
I don't know scala well enough to be totally sure this second argument will be lazily evaluated. I'm assuming it will be but just want to point it out in case there's some special syntax that might need to be used to get that behavior